src/dataflow/src/render/reduce.rs

```diff
@@ -87,20 +87,39 @@ where
             // Our first step is to extract `(key, vals)` from `input`.
             // We do this carefully, attempting to avoid unneccesary allocations
             // that would result from cloning rows in input arrangements.
-            let group_key_clone = group_key.clone();
-            let aggregates_clone = aggregates.clone();
-
-            // Tracks the required number of columns to extract.
-            let mut columns_needed = 0;
-            for key in group_key.iter() {
-                for column in key.support() {
-                    columns_needed = std::cmp::max(columns_needed, column + 1);
-                }
+
+            let input_arity = input.arity();
+
+            // Form an operator for evaluating key expressions.
+            let mut key_mfp = expr::MapFilterProject::new(input_arity)
+                .map(group_key.iter().cloned())
+                .project(input_arity..input_arity + group_key.len());
```

Comment:
maybe this is just dumb-ness on my part but the associativity of the range operator was extremely non-obvious to me at first glance. I kept parsing this as

```
(input_arity..input_arity) + group_key.len()
```
I'm not sure if theres an easy way to make that more clear or if this is just me being stupid

Reply:
I'm not sure if there is a way. `cargo fmt` insists on removing the space around `..` and introducing it at `+`, which is how I deal with this in non `cargo fmt` code. I can put some parentheses in place and we'll see if clippy says that they are wrong and I should feel bad about myself.

Reply:
`cargo check` doesn't complain, so I'll ship that with the next doc-oriented commit and we'll see what happens.

Reply:
you can also run `bin/pre-push` on your laptop to see what clippy will say fwiw - clippy seemed fine with a paren around the + on my laptop